File: src/main/java/com/chidozie/core/renderEngine/OBJFileLoader.java
package com.chidozie.core.renderEngine;

import net.adventuregame.models.RawModel;
import org.joml.Vector2f;
import org.joml.Vector3f;

import java.io.*;
import java.util.ArrayList;
import java.util.List;

public class OBJFileLoader {

	public static RawModel loadOBJ(String objFileName, Loader loader) {
		InputStream isr = null;
		try {
			isr = OBJFileLoader.class.getClassLoader().getResourceAsStream("assets/adventuregame/models/" + objFileName + ".obj");
			if (isr == null) {
				System.err.println("File not found in res; don't use any extension");
				throw new FileNotFoundException("File not found: " + objFileName + ".obj");
			}
		} catch (FileNotFoundException e) {
			System.err.println("File not found in res; don't use any extension");
			e.printStackTrace();
			System.exit(-1);
		}
		BufferedReader reader = new BufferedReader(new InputStreamReader(isr));
		String line;
		List<Vertex> vertices = new ArrayList<Vertex>();
		List<Vector2f> textures = new ArrayList<Vector2f>();
		List<Vector3f> normals = new ArrayList<Vector3f>();
		List<Integer> indices = new ArrayList<Integer>();
		try {
			while (true) {
				line = reader.readLine();
				if (line.startsWith("v ")) {
					String[] currentLine = line.split(" ");
					Vector3f vertex = new Vector3f(Float.parseFloat(currentLine[1]),
							Float.parseFloat(currentLine[2]),
							Float.parseFloat(currentLine[3]));
					Vertex newVertex = new Vertex(vertices.size(), vertex);
					vertices.add(newVertex);

				} else if (line.startsWith("vt ")) {
					String[] currentLine = line.split(" ");
					Vector2f texture = new Vector2f(Float.parseFloat(currentLine[1]),
							Float.parseFloat(currentLine[2]));
					textures.add(texture);
				} else if (line.startsWith("vn ")) {
					String[] currentLine = line.split(" ");
					Vector3f normal = new Vector3f(Float.parseFloat(currentLine[1]),
							Float.parseFloat(currentLine[2]),
							Float.parseFloat(currentLine[3]));
					normals.add(normal);
				} else if (line.startsWith("f ")) {
					break;
				}
			}
			while (line != null && line.startsWith("f ")) {
				String[] currentLine = line.split(" ");
				String[] vertex1 = currentLine[1].split("/");
				String[] vertex2 = currentLine[2].split("/");
				String[] vertex3 = currentLine[3].split("/");
				processVertex(vertex1, vertices, indices);
				processVertex(vertex2, vertices, indices);
				processVertex(vertex3, vertices, indices);
				line = reader.readLine();
			}
			reader.close();
		} catch (IOException e) {
			System.err.println("Error reading the file");
			e.printStackTrace();
			System.exit(-1);
		}
		removeUnusedVertices(vertices);
		float[] verticesArray = new float[vertices.size() * 3];
		float[] texturesArray = new float[vertices.size() * 2];
		float[] normalsArray = new float[vertices.size() * 3];
		float furthest = convertDataToArrays(vertices, textures, normals, verticesArray, texturesArray, normalsArray);
		int[] indicesArray = convertIndicesListToArray(indices);
		ModelData data = new ModelData(verticesArray, texturesArray, normalsArray, indicesArray, furthest);
		return loader.loadToVAO(data.getVertices(), data.getTextureCoords(), data.getNormals(), data.getIndices());
	}

	private static void processVertex(String[] vertex, List<Vertex> vertices, List<Integer> indices) {
		int index = Integer.parseInt(vertex[0]) - 1;
		Vertex currentVertex = vertices.get(index);
		int textureIndex = Integer.parseInt(vertex[1]) - 1;
		int normalIndex = Integer.parseInt(vertex[2]) - 1;
		if (!currentVertex.isSet()) {
			currentVertex.setTextureIndex(textureIndex);
			currentVertex.setNormalIndex(normalIndex);
			indices.add(index);
		} else {
			dealWithAlreadyProcessedVertex(currentVertex, textureIndex, normalIndex, indices,
					vertices);
		}
	}

	private static int[] convertIndicesListToArray(List<Integer> indices) {
		int[] indicesArray = new int[indices.size()];
		for (int i = 0; i < indicesArray.length; i++) {
			indicesArray[i] = indices.get(i);
		}
		return indicesArray;
	}

	private static float convertDataToArrays(List<Vertex> vertices, List<Vector2f> textures,
			List<Vector3f> normals, float[] verticesArray, float[] texturesArray,
			float[] normalsArray) {
		float furthestPoint = 0;
		for (int i = 0; i < vertices.size(); i++) {
			Vertex currentVertex = vertices.get(i);
			if (currentVertex.getLength() > furthestPoint) {
				furthestPoint = currentVertex.getLength();
			}
			Vector3f position = currentVertex.getPosition();
			Vector2f textureCoord = textures.get(currentVertex.getTextureIndex());
			Vector3f normalVector = normals.get(currentVertex.getNormalIndex());
			verticesArray[i * 3] = position.x;
			verticesArray[i * 3 + 1] = position.y;
			verticesArray[i * 3 + 2] = position.z;
			texturesArray[i * 2] = textureCoord.x;
			texturesArray[i * 2 + 1] = 1 - textureCoord.y;
			normalsArray[i * 3] = normalVector.x;
			normalsArray[i * 3 + 1] = normalVector.y;
			normalsArray[i * 3 + 2] = normalVector.z;
		}
		return furthestPoint;
	}

	private static void dealWithAlreadyProcessedVertex(Vertex previousVertex, int newTextureIndex,
			int newNormalIndex, List<Integer> indices, List<Vertex> vertices) {
		if (previousVertex.hasSameTextureAndNormal(newTextureIndex, newNormalIndex)) {
			indices.add(previousVertex.getIndex());
		} else {
			Vertex anotherVertex = previousVertex.getDuplicateVertex();
			if (anotherVertex != null) {
				dealWithAlreadyProcessedVertex(anotherVertex, newTextureIndex, newNormalIndex,
						indices, vertices);
			} else {
				Vertex duplicateVertex = new Vertex(vertices.size(), previousVertex.getPosition());
				duplicateVertex.setTextureIndex(newTextureIndex);
				duplicateVertex.setNormalIndex(newNormalIndex);
				previousVertex.setDuplicateVertex(duplicateVertex);
				vertices.add(duplicateVertex);
				indices.add(duplicateVertex.getIndex());
			}

		}
	}
	
	private static void removeUnusedVertices(List<Vertex> vertices){
		for(Vertex vertex:vertices){
			if(!vertex.isSet()){
				vertex.setTextureIndex(0);
				vertex.setNormalIndex(0);
			}
		}
	}

}

File: src/main/kotlin/com/adv/core/renderEngine/OBJFileLoader.kt
package com.adv.core.renderEngine

import net.adventuregame.models.RawModel
import org.joml.Vector2f
import org.joml.Vector3f
import java.io.*

object OBJFileLoader {
    fun loadOBJ(objFileName: String, loader: Loader): RawModel? {
        val path = "assets/adventuregame/models/$objFileName.obj"
        val isr = OBJFileLoader::class.java.classLoader.getResourceAsStream(path)
            ?: throw FileNotFoundException("File not found in res: $path (don’t use extension)")

        val reader = BufferedReader(InputStreamReader(isr))
        val vertices = mutableListOf<Vertex>()
        val textures = mutableListOf<Vector2f>()
        val normals = mutableListOf<Vector3f>()
        val indices = mutableListOf<Int>()

        var line: String? = reader.readLine()

        try {
            while (line != null && !line.startsWith("f ")) {
                val tokens = line.trim().split("\\s+".toRegex())

                when {
                    line.startsWith("v ") -> {
                        val vertex = Vector3f(tokens[1].toFloat(), tokens[2].toFloat(), tokens[3].toFloat())
                        vertices.add(Vertex(vertices.size, vertex))
                    }
                    line.startsWith("vt ") -> {
                        val texture = Vector2f(tokens[1].toFloat(), tokens[2].toFloat())
                        textures.add(texture)
                    }
                    line.startsWith("vn ") -> {
                        val normal = Vector3f(tokens[1].toFloat(), tokens[2].toFloat(), tokens[3].toFloat())
                        normals.add(normal)
                    }
                }

                line = reader.readLine()
            }

            while (line != null && line.startsWith("f ")) {
                val tokens = line.trim().split("\\s+".toRegex())

                val verticesData = listOf(
                    tokens[1].split("/"),
                    tokens[2].split("/"),
                    tokens[3].split("/")
                )

                verticesData.forEach { processVertex(it.toTypedArray(), vertices, indices) }
                line = reader.readLine()
            }

            reader.close()
        } catch (e: IOException) {
            System.err.println("Error reading the OBJ file: ${e.message}")
            e.printStackTrace()
            return null
        }

        removeUnusedVertices(vertices)

        val verticesArray = FloatArray(vertices.size * 3)
        val texturesArray = FloatArray(vertices.size * 2)
        val normalsArray = FloatArray(vertices.size * 3)
        val furthest = convertDataToArrays(vertices, textures, normals, verticesArray, texturesArray, normalsArray)
        val indicesArray = convertIndicesListToArray(indices)

        val data = ModelData(verticesArray, texturesArray, normalsArray, indicesArray, furthest)
        return loader.loadToVAO(data.vertices!!, data.textureCoords!!, data.normals!!, data.indices!!)
    }


    private fun processVertex(vertex: Array<String?>, vertices: MutableList<Vertex>, indices: MutableList<Int>) {
        val index = vertex[0]!!.toInt() - 1
        val currentVertex = vertices.get(index)
        val textureIndex = vertex[1]!!.toInt() - 1
        val normalIndex = vertex[2]!!.toInt() - 1
        if (!currentVertex.isSet) {
            currentVertex.textureIndex = textureIndex
            currentVertex.normalIndex = normalIndex
            indices.add(index)
        } else {
            dealWithAlreadyProcessedVertex(
                currentVertex, textureIndex, normalIndex, indices,
                vertices
            )
        }
    }

    private fun convertIndicesListToArray(indices: MutableList<Int>): IntArray {
        val indicesArray = IntArray(indices.size)
        for (i in indicesArray.indices) {
            indicesArray[i] = indices[i]
        }
        return indicesArray
    }

    private fun convertDataToArrays(
        vertices: MutableList<Vertex>, textures: MutableList<Vector2f>,
        normals: MutableList<Vector3f>, verticesArray: FloatArray, texturesArray: FloatArray,
        normalsArray: FloatArray
    ): Float {
        var furthestPoint = 0f
        for (i in vertices.indices) {
            val currentVertex = vertices[i]
            if (currentVertex.length > furthestPoint) {
                furthestPoint = currentVertex.length
            }
            val position = currentVertex.position!!
            val textureCoord = textures[currentVertex.textureIndex]
            val normalVector = normals[currentVertex.normalIndex]
            verticesArray[i * 3] = position.x
            verticesArray[i * 3 + 1] = position.y
            verticesArray[i * 3 + 2] = position.z
            texturesArray[i * 2] = textureCoord.x
            texturesArray[i * 2 + 1] = 1 - textureCoord.y
            normalsArray[i * 3] = normalVector.x
            normalsArray[i * 3 + 1] = normalVector.y
            normalsArray[i * 3 + 2] = normalVector.z
        }
        return furthestPoint
    }

    private fun dealWithAlreadyProcessedVertex(
        previousVertex: Vertex, newTextureIndex: Int,
        newNormalIndex: Int, indices: MutableList<Int>, vertices: MutableList<Vertex>
    ) {
        if (previousVertex.hasSameTextureAndNormal(newTextureIndex, newNormalIndex)) {
            indices.add(previousVertex.index)
        } else {
            val anotherVertex = previousVertex.duplicateVertex
            if (anotherVertex != null) {
                dealWithAlreadyProcessedVertex(
                    anotherVertex, newTextureIndex, newNormalIndex,
                    indices, vertices
                )
            } else {
                val duplicateVertex = Vertex(vertices.size, previousVertex.position!!)
                duplicateVertex.textureIndex = newTextureIndex
                duplicateVertex.normalIndex = newNormalIndex
                previousVertex.duplicateVertex = duplicateVertex
                vertices.add(duplicateVertex)
                indices.add(duplicateVertex.index)
            }
        }
    }

    private fun removeUnusedVertices(vertices: MutableList<Vertex>) {
        for (vertex in vertices) {
            if (!vertex.isSet) {
                vertex.textureIndex = 0
                vertex.normalIndex = 0
            }
        }
    }
}